src/sys/unix/uds/datagram.rs

```diff
@@ -129,6 +129,29 @@ impl UnixDatagram {
     pub(crate) fn shutdown(&self, how: Shutdown) -> io::Result<()> {
         self.inner.shutdown(how)
     }
+
+    pub(crate) fn unlink(mut self) {
```

Comment:
I have a comment about naming in this method, but I think it could be structured differently. In situations where there is a `None => None` and (or) `Err(_) => None`, `map` is probably going to making the branching read a little nicer.

How about something like:
```
let socket = mem::replace(&mut self.inner, unsafe {
    net::UnixListener::from_raw_fd(-1)
});
let _ = socket.local_addr().map(|addr| {
    addr.as_pathname().map(|path| {
        let path = path.as_os_str().to_owned();
        let _ = std::fs::remove_file(path);
    })
});
mem::forget(self);
```

I think `match` and `if let` can certainly improve readability, but in this case there is no action to take on the `None` or `Err` variants.

This applies to the `UnixListener` method as well.

Reply:
The problem is that the socket has to be dropped first before we can unlink. Also I realized that `remove_file` doesn't work because it remove unconditionally the socket file, which is not what we want. I'll add a test case.

Reply:
Nevermind, see my comment below.